executor/snapshot.h

```diff
@@ -0,0 +1,226 @@
+	output_size = static_cast<uint64>(rpc::Const::MaxOutputSize) - sizeof(rpc::SnapshotHeaderT);
+}
+
+static void SnapshotSetup(char** argv, int argc)
+{
+	flag_snapshot = true;
+	// This allows to see debug output during early setup.
+	// If debug is not actually enabled, it will be turned off in parse_handshake.
+	flag_debug = true;
+#if GOOS_linux
+	// In snapshot mode executor output is redirected to /dev/kmsg.
+	// This is required to turn off rate limiting of writes.
+	write_file("/proc/sys/kernel/printk_devkmsg", "on\n");
+#endif
+	FindIvshmemDevices();
+	// Wait for the host to write handshake_req into input memory.
+	while (ivs.hdr->state != rpc::SnapshotState::Handshake)
+		sleep_ms(10);
+	const auto& req = *static_cast<handshake_req*>(ivs.input);
+	parse_handshake(req);
+	for (const auto& feat : features) {
+		if (!(req.features & feat.id))
+			continue;
+		debug("setting up feature %s\n", rpc::EnumNameFeature(feat.id));
+		const char* reason = feat.setup();
+		if (reason)
+			failmsg("feature setup failed", "reason: %s", reason);
+	}
+}
+
+constexpr size_t kOutputPopulate = 256 << 10;
+constexpr size_t kInputPopulate = 64 << 10;
+constexpr size_t kGlobalsPopulate = 4 << 10;
+constexpr size_t kDataPopulate = 8 << 10;
+constexpr size_t kCoveragePopulate = 32 << 10;
+constexpr size_t kThreadsPopulate = 2;
+
+static void SnapshotSetState(rpc::SnapshotState state)
+{
+	debug("changing stapshot state %s -> %s\n",
+	      rpc::EnumNameSnapshotState(ivs.hdr->state), rpc::EnumNameSnapshotState(state));
+	std::atomic_signal_fence(std::memory_order_seq_cst);
+	ivs.hdr->state = state;
+	// The register contains VM index shifted by 16 (the host part is VM index 1)
+	// + interrup vector index (0 in our case).
+	*ivs.doorbell = 1 << 16;
+}
+
+// PopulateMemory prefaults anon memory (we want to avoid minor page faults as well).
+static void PopulateMemory(void* ptr, size_t size)
+{
+	ptr = (void*)(uintptr_t(ptr) & ~(getpagesize() - 1));
+	if (madvise(ptr, size, MADV_POPULATE_WRITE))
+		failmsg("populate madvise failed", "ptr=%p size=%zu", ptr, size);
+}
+
+// TouchMemory prefaults non-anon shared memory.
+static void TouchMemory(void* ptr, size_t size)
+{
+	size_t const kPageSize = getpagesize();
+	for (size_t i = 0; i < size; i += kPageSize)
+		(void)((volatile char*)ptr)[i];
+}
+
+#if SYZ_EXECUTOR_USES_FORK_SERVER
+static void SnapshotPrepareParent()
+{
+	TouchMemory((char*)output_data + output_size - kOutputPopulate, kOutputPopulate);
+	// Notify SnapshotStart that we finished prefaulting memory in the parent.
+	output_data->completed = 1;
+	// Wait for the request to come, so that we give it full time slice to execute.
+	// This process will start waiting for the child as soon as we return.
+	while (ivs.hdr->state != rpc::SnapshotState::Execute)
+		;
+}
+#endif
+
+static void SnapshotStart()
+{
+	debug("SnapshotStart\n");
+	// Prefault as much memory as we can before the snapshot is taken.
```

Comment:
Please add a comment re why we should do this.

Reply:
Done.